File: src/LLMChain.js
import { ChatPromptTemplate } from '@langchain/core/prompts';
import { StringOutputParser } from '@langchain/core/output_parsers';
import { parseDiff, applyDiff } from './diffUtils';

class LLMChain {
  constructor(llm, messageHandler, backend, options) {
    this.llm = llm;
    this.messageHandler = messageHandler;
    this.backend = backend;
    this.options = options;
    this.prompt = ChatPromptTemplate.fromTemplate(
      `You are a helpful AI assistant that helps with coding.

Here is the current codebase:
{codebase}

Here is the current diff:
{diff}

Here is the output of the last command:
{lastCommandOutput}

Here is the user's query:
{input}`
    );
    this.outputParser = new StringOutputParser();
    this.chain = this.prompt.pipe(this.llm).pipe(this.outputParser);
  }

  async handleQuery(query, codebase, diff) {
    try {
      const messages = this.messageHandler.getMessages();
      const lastCommandOutput = messages.length > 0 ? messages[messages.length - 1].text : '';
      const response = await this.chain.invoke({
        input: query,
        chat_history: messages,
        codebase,
        diff,
        lastCommandOutput,
      });
      this.messageHandler.addMessage({ sender: 'ai', text: response });

      const parsedDiff = parseDiff(response);
      if (parsedDiff) {
        try {
          await applyDiff(parsedDiff);
          const newDiff = await this.backend.diff();
          this.messageHandler.addMessage({ sender: 'system', text: 'Diff applied successfully.' });
          if (this.options.autoCommit) {
            await this.backend.record('Auto-commit');
            this.messageHandler.addMessage({ sender: 'system', text: 'Changes auto-committed.' });
          }
          return newDiff;
        } catch (error) {
          this.messageHandler.addMessage({
            sender: 'system',
            text: `Error applying diff: ${error.message}`,
          });
        }
      }
    } catch (error) {
      this.messageHandler.addMessage({
        sender: 'system',
        text: `Error invoking LLM: ${error.message}`,
      });
    }
    return diff;
  }
}

export default LLMChain;

File: java/src/main/java/com/pijul/aider/llm/LLMChain.java
package com.pijul.aider.llm;

import com.pijul.aider.Container;
import dev.langchain4j.googleai.gemini.ChatPromptTemplate;
import dev.langchain4j.googleai.gemini.LLM;
import dev.langchain4j.googleai.gemini.LLMChain;
import dev.langchain4j.googleai.gemini.output.OutputParser;
import dev.langchain4j.googleai.gemini.prompts.PromptTemplate;
import dev.langchain4j.googleai.gemini.output_parsers.StringOutputParser;
import java.util.List;
import java.util.concurrent.CompletableFuture;

public class LLMChain {
    private final LLMChain chain;
    private final OutputParser<String> outputParser;
    private final Container container;

    public LLMChain(LLM llm, Container container) {
        this.container = container;
        PromptTemplate prompt = ChatPromptTemplate.fromTemplate(
            "You are a helpful AI assistant that helps with coding.\n\n" +
            "Here is the current codebase: {codebase}\n\n" +
            "Here is the current diff: {diff}\n\n" +
            "Here is the output of the last command: {lastCommandOutput}\n\n" +
            "Here is the user's query: {input}"
        );
        
        this.chain = new LLMChain(prompt, llm, new StringOutputParser());
        this.outputParser = new StringOutputParser();
    }

    public CompletableFuture<String> handleQuery(String query, String codebase, String diff, String lastCommandOutput) {
        return chain.invoke(List.of(
            new PromptTemplate.Input("codebase", codebase),
            new PromptTemplate.Input("diff", diff),
            new PromptTemplate.Input("lastCommandOutput", lastCommandOutput),
            new PromptTemplate.Input("input", query)
        )).thenApply(output -> {
            System.out.println("AI Response: " + output);
            container.addMessage("ai", "Response: " + output);
            
            // Apply diff if present in response
            if (output.contains("diff")) {
                // Implement diff application logic here
                System.out.println("Applying diff from AI response...");
                // TODO: Implement actual diff application
            }
            return output;
        });
    }
}